bundles/org.openhab.core.semantics/src/main/java/org/openhab/core/semantics/internal/CustomTagRegistryImpl.java

```diff
@@ -0,0 +1,146 @@
+        createSemanticTag(customTag.getUID(), customTag.getLabel(), customTag.getDescription(),
+                customTag.getSynonyms());
+    }
+
+    @Override
+    protected void onRemoveElement(CustomTag customTag) {
+        logger.trace("onRemoveElement {}", customTag.getUID());
+        super.onRemoveElement(customTag);
+        Class<? extends Tag> tag = SemanticTags.getById(customTag.getUID());
+        if (tag != null) {
+            SemanticTags.remove(tag);
+        }
+    }
+
+    @Override
+    protected void onUpdateElement(CustomTag oldCustomTag, CustomTag customTag) throws IllegalArgumentException {
+        logger.trace("onUpdateElement {}", customTag.getUID());
+        super.onUpdateElement(oldCustomTag, customTag);
+        Class<? extends Tag> tag = SemanticTags.getById(customTag.getUID());
+        if (tag == null) {
+            throw new IllegalArgumentException("No existing tag with id " + customTag.getUID());
+        }
+        SemanticTags.remove(tag);
+        createSemanticTag(customTag.getUID(), customTag.getLabel(), customTag.getDescription(),
+                customTag.getSynonyms());
+    }
+
+    private void createSemanticTag(String uid, String label, String description, List<String> synonyms)
+            throws IllegalArgumentException {
+        Class<? extends Tag> tag = SemanticTags.getById(uid);
+        if (tag != null) {
+            throw new IllegalArgumentException("Tag " + tag.getAnnotation(TagInfo.class).id() + " already exist");
+        }
+        Class<? extends Tag> parentTag = null;
+        int lastSeparator = uid.lastIndexOf("_");
+        if (lastSeparator <= 0) {
+            throw new IllegalArgumentException("Invalid tag id " + uid);
+        }
+        String name = uid.substring(lastSeparator + 1);
+        parentTag = SemanticTags.getById(uid.substring(0, lastSeparator));
+        if (parentTag == null) {
+            throw new IllegalArgumentException("No existing parent tag with id " + uid.substring(0, lastSeparator));
+        } else if (!name.matches("[A-Z][a-zA-Z0-9]+")) {
+            throw new IllegalArgumentException("Invalid tag name " + name);
+        }
+        tag = SemanticTags.getById(name);
+        if (tag != null) {
+            throw new IllegalArgumentException("Tag " + tag.getAnnotation(TagInfo.class).id() + " already exist");
+        }
+        if (SemanticTags.add(name, parentTag, label, String.join(",", synonyms), description) == null) {
+            throw new IllegalArgumentException("Failed to create semantic tag " + uid);
+        }
+    }
+
+    @Override
+    @Reference
+    protected void setReadyService(ReadyService readyService) {
+        super.setReadyService(readyService);
+    }
+
+    @Override
+    protected void unsetReadyService(ReadyService readyService) {
+        super.unsetReadyService(readyService);
+    }
+
+    @Reference(cardinality = ReferenceCardinality.OPTIONAL, policy = ReferencePolicy.DYNAMIC)
```

Comment:
I would suggest to use constructor injection and make the `ManagedProvider` mandatory. See the discussion in https://github.com/openhab/openhab-core/issues/3017.

Reply:
Ok